instrumentation/spring/spring-integration-4.1/library/src/main/java/io/opentelemetry/instrumentation/spring/integration/MessageSpanLinkExtractor.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.spring.integration;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanLinkExtractor;
+
+final class MessageSpanLinkExtractor implements SpanLinkExtractor<MessageWithChannel> {
+  private final SpanLinkExtractor<MessageWithChannel> delegate;
+
+  MessageSpanLinkExtractor(SpanLinkExtractor<MessageWithChannel> delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public SpanContext extract(Context parentContext, MessageWithChannel messageWithChannel) {
+    SpanContext spanFromMessage = delegate.extract(parentContext, messageWithChannel);
+    SpanContext parentSpan = Span.fromContext(parentContext).getSpanContext();
+    if (referencesSameSpan(spanFromMessage, parentSpan)) {
+      return SpanContext.getInvalid();
+    }
+    return spanFromMessage;
+  }
+
+  // SpanContext#equals() includes e.g. remote flag, which we don't really care about here
+  // we just want to avoid adding links to spans with the same id, flags don't matter at all
```

Comment:
Is it normal situation when `referencesSameSpan==true` but `spanFromMessage.equals(parentSpan) == false`? I am not sure that we should work around such situation...

Reply:
`SpanContext.equals()` also includes the remote flag, which is always set to true when you extract the context from the message. So in a completely "local" scenario, when you have one message channel sending a message to another you'd end up with a link to the parent span in the second channel's span.